### PR TITLE
[ci][7up/2] upgrade ci tests to py39

### PR DESCRIPTION
**Pull request by can-anyscale**
Upgrade all CI tests to run with py39 by default. Only upgrade python, no other dependencies are upgraded.

Test:
- CI

### CODE REVIEW

File: python/ray/tests/test_client_warnings.py [hunk 1]
```diff
@@ -18,20 +18,19 @@ class LoggerSuite(unittest.TestCase):
     """Test client warnings are raised when many tasks are scheduled"""
 
     def testOutboundMessageSizeWarning(self):
-        with ray_start_client_server() as ray:
+        with self.assertWarns(UserWarning) as cm, ray_start_client_server() as ray:
```
Comment (can-anyscale) [inline python/ray/tests/test_client_warnings.py:21]: pytest hits a bug when `assertWarns` context is nested; I rewrote to avoid that bug while keeping the same test semantic

File: python/ray/tests/test_runtime_env_complicated.py [hunk 2]
```diff
@@ -870,16 +870,16 @@ def test(self):
...
 
-    # pip requirements file from Ray Summit 2021 demo.
+    # pip requirements file from Ray Summit 2021 demo; updated to be compatible with
+    # recent python versions
     requirement_path = tmp_path / "requirements.txt"
     requirement_path.write_text(
         "\n".join(
             [
                 "ray[serve, tune]",
-                "texthero",
```
Comment (aslonnie) [inline python/ray/tests/test_runtime_env_complicated.py:879]: why is this removed?

File: python/ray/tests/test_runtime_env_complicated.py [hunk 3]
```diff
@@ -870,16 +870,16 @@ def test(self):
...
                 "PyGithub",
                 "xgboost_ray",  # has Ray as a dependency
-                "pandas==1.1",  # pandas 1.2.4 in the demo, but not supported on py36
+                "pandas==1.5.3",  # pandas 1.2.4 in the demo, but not supported on py39
```
Comment (aslonnie) [inline python/ray/tests/test_runtime_env_complicated.py:882]: maybe just remove the comment?

File: python/ray/tests/test_client_warnings.py (hunk 1 above)
Comment (aslonnie) [reply python/ray/tests/test_client_warnings.py:21]: is it possible to split this small part into a separate PR?

File: ci/ray_ci/windows/build_base.sh [hunk 4]
```diff
@@ -3,6 +3,8 @@
 set -ex
 
 conda init 
+# newer requests version is needed for python 3.9+
+conda install -q -y python="${PYTHON}" requests=2.31.0
```
Comment (aslonnie) [inline ci/ray_ci/windows/build_base.sh:7]: the `requests` package should be installed via pip rather than via conda?

could you provide more context on why "requests" version is needed?

File: ci/ray_ci/test_ray_docker_container.py [hunk 5]
```diff
@@ -24,16 +24,16 @@ def _mock_run_script(input: List[str]) -> None:
             "ci.ray_ci.docker_container.LinuxContainer.run_script",
             side_effect=_mock_run_script,
         ):
-            container = RayDockerContainer("3.8", "cu11.8.0", "ray")
+            container = RayDockerContainer("3.9", "cu11.8.0", "ray")
```
Comment (aslonnie) [inline ci/ray_ci/test_ray_docker_container.py:27]: maybe just changes these to a fake version number? like "3.99"? so that we won't need to update these tests in the future?

File: ci/ray_ci/docker_container.py [hunk 6]
```diff
@@ -73,7 +73,7 @@ def _get_canonical_tag(self) -> str:
         return self.canonical_tag if self.canonical_tag else self._get_image_tags()[0]
 
     def get_python_version_tag(self) -> str:
-        return f"-py{self.python_version.replace('.', '')}"  # 3.8 -> py38
+        return f"-py{self.python_version.replace('.', '')}"  # 3.9 -> py39
```
Comment (aslonnie) [inline ci/ray_ci/docker_container.py:76]: change the comment to `"3.x -> 3x"` maybe?

File: ci/ray_ci/docker_container.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
     "cu12.1.1",
 ]
 GPU_PLATFORM = "cu11.8.0"
-DEFAULT_PYTHON_VERSION = "3.8"
+DEFAULT_PYTHON_VERSION = "3.9"
```
Comment (aslonnie) [inline ci/ray_ci/docker_container.py:17]: why is this declared twice here and in `builder_container.py`, and also `builder.py`? can we unify?

**Review by aslonnie — COMMENTED** [review]
my high level idea is that we should try to minimize the required changes for the next time.

also, if this goes in, will python 3.11 ml image build?

**Comment by can-anyscale** [comment]
Yess, let me break this down into smaller pieces.

These changes are not enough to support ml-311 yet.

File: python/ray/tests/test_runtime_env_complicated.py (hunk 2 above)
Comment (can-anyscale) [reply python/ray/tests/test_runtime_env_complicated.py:879]: texthero doesn't have py39 version

File: ci/ray_ci/windows/build_base.sh (hunk 4 above)
Comment (can-anyscale) [reply ci/ray_ci/windows/build_base.sh:7]: conda itself is broken without upgrading requests https://buildkite.com/ray-project/premerge/builds/15486#018cad7e-92d5-4d58-81d1-3c100afbd230/12648-12671

**Comment by can-anyscale** [comment]
Tests are all green \o/, do you want to go forward @jjyao , @rkooo567 , @rickyyx , @matthewdeng , @edoakes , @scottjlee , thankks

**Review by aslonnie — APPROVED** [review]
looks much cleaner now.

File: python/ray/_private/runtime_env/utils.py [hunk 8]
```diff
@@ -89,6 +89,10 @@ async def check_output_cmd(
         #     https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.stderr
         #   * Avoid mixing multiple outputs of concurrent cmds.
         stdout, _ = await proc.communicate()
+    except asyncio.exceptions.CancelledError as e:
```
Comment (rickyyx) [inline python/ray/_private/runtime_env/utils.py:92]: do we have any tests/repro steps to provide signals for this change?

Comment (can-anyscale) [reply python/ray/_private/runtime_env/utils.py:92]: yess, I updated this because one of those test will fail without this in py 3.9. I forgot which one, let me remove this change and trace back which test needs this.

File: python/ray/_private/runtime_env/utils.py [hunk 9]
```diff
@@ -89,6 +89,10 @@ async def check_output_cmd(
...
+        # since Python 3.9, when cancelled, the inner process needs to throw as it is
+        # for asyncio to timeout properly https://bugs.python.org/issue40607
```
Comment (jjyao) [inline python/ray/_private/runtime_env/utils.py:94]: What's the issue here? I checked the bug and the fix (https://github.com/python/cpython/pull/20054/files). Seems we can catch CancelledError and raise a different one.

**Review by edoakes — APPROVED** [review]

File: python/ray/_private/runtime_env/utils.py (hunk 9 above)
Comment (can-anyscale) [reply python/ray/_private/runtime_env/utils.py:94]: The issue is that this code throws as RuntimeError instead of CancelledError. Before https://github.com/python/cpython/commit/382a5635bd10c237c3e23e346b21cde27e48d7fa, regardless of that error this code throw, asyncio/tasks.py will throw TimeoutError which is what we need. After the change, asyncio/tasks.py will throw RuntimeError up in the line


```
except exceptions.CancelledError as exc:
                raise exceptions.TimeoutError() from exc
```

Comment (jjyao) [reply python/ray/_private/runtime_env/utils.py:94]: I see.

**Review by jjyao — APPROVED** [review]

File: python/ray/_private/runtime_env/utils.py (hunk 8 above)
Comment (can-anyscale) [reply python/ray/_private/runtime_env/utils.py:92]: @rickyyx this is the test that failed https://buildkite.com/ray-project/postmerge/builds/2431#018ceade-2b21-47b5-b554-1494daed53fe/6-1061; the runtime env is expected to timed out but is thrown as runtimeerror instead.

**Comment by can-anyscale** [comment]
@matthewdeng , @scottjlee , @raulchen let me know if you have strong opinion about not running tests using python 3.9 in CI, thankkks.
